### DISCUSSION TITLE
Project: PyDM - Python Display Manager

### DISCUSSION THREAD

**Discussion by klauer** [pcdshub/epics-codeathon-2022-python#2] [Projects]
Source code repository: https://github.com/slaclab/pydm/
Documentation: http://slaclab.github.io/pydm/

![image](https://user-images.githubusercontent.com/5139267/164513700-90efc770-7619-4b71-b573-8f678c168127.png)

**Comment by klauer** [comment]
## Possible project
* Finish up PyDM 2.0 upgrade, a big work-in-progress PR 
* Rework data source plugin system
* Add PVAccess support

## Issue / Pull Request
~https://github.com/slaclab/pydm/pull/621~ Package maintainers indicate that a different effort may be in the works. Not clear if this will be feasible for the codeathon.

**Comment by klauer** [comment]
## Possible project
* Finish up PR to enhance designer experience

## Issue / Pull Request
https://github.com/slaclab/pydm/pull/795

**Reply by klauer** [reply]
Small update: closer to merging this. 🎆

**Comment by klauer** [comment]
## Possible project

EDM -> PyDM converter tool

## Issue / Pull Request

* Related but not for EDM: https://github.com/slaclab/pydm/pull/723
* MEDM (.adl file) converter: https://github.com/BCDA-APS/adl2pydm/

**Comment by klauer** [comment]
## Project idea

General issue:
* Introduce pydm to the community
* Sort through the issue list https://github.com/slaclab/pydm/issues
* Reproduce the issue and make an attempt at a fix

## Pull Request / Issue link

Any open issue:
https://github.com/slaclab/pydm/issues

**Comment by klauer** [comment]
## Project idea

* GUI scaling?
* caQtDM has a solution for this; could we provide useful information to the PyDM developers about this?

**Reply by prjemian** [reply]
Commented on the issue for this group today: https://github.com/slaclab/pydm/issues/574#issuecomment-1121347612

**Comment by klauer** [comment]
## Project idea

* Work on entrypoints support
* For external tools
* For data plugins

**Reply by klauer** [reply]
PR https://github.com/slaclab/pydm/pull/844

**Comment by ZLLentz** [comment]
Startup speed investigation: case study `pmps-ui`

Goals:
- Create execution profiles of `pmps-ui`'s various tabs
- Analyze the slowest tabs and document why they are slow
- Figure out if there's anything we can work on in PyDM or if there are any best practices to document for PyDM users

**Reply by ZLLentz** [reply]
Progress:

- Created a script for launching single tabs in order to profile them
- Ran profiles on all the tabs
- unsurprisingly, `fast_faults` and `preemptive_requests` are the slow tabs, but they are slow in different ways
- `fast_faults` spends most of its execution time in the evaluation of PyDM rules. The display opens quickly to a blank page, then takes away the user's control to evaluate rules and render, then gives control back eventually.
- `preemptive_requests` spends most of its time opening/instantiating/rendering the various repeated widgets. The display takes a while to to open, but once it opens it is immediately responsive.
- the app loads great when both of these are disabled

Next: investigate the rules execution. How does this work? Why does the user lose control here? Are there any lessons we can learn or improvements to make here?

**Reply by ZLLentz** [reply]
The rules engine is a QThread that runs in a loop at 30Hz, evaluating all the rules each time. This includes the first loop which needs to do a lot of heavy lifting for a screen with as many rules as pmps-ui. I think this is why the initialization of the fast_faults tab freezes the gui in place after opening.

https://github.com/slaclab/pydm/blob/f60a7ac23826f25b90ebf303c56f168feb7c7751/pydm/widgets/rules.py#L220-L230

**Reply by ZLLentz** [reply]
I updated some of our internal profiling tools to help investigate this.
It seems like about half the time comes from the rules engine and half the time comes from loading the files and creating the widgets. I tried a patch to fix the file loading which helped the file loading but had unfortunate consequences on the rules evaluation, resulting in a net loss. I'll probably put this up as a draft PR for discussion as a concept.

I think there are a few best practices takeaways to make here:
1. Creating and showing ~500-1000 widgets dynamically has a sharp performance penalty. For applications like this it's worth investigating pagination and slow background loading that won't impact the main thread- it's important to keep the event queue small or the display will feel clunky and unresponsive.
2. The rules engine is a useful tool for designer, but there is no reason to rely on it for behavior in the context of a procedurally generated application that is launched from a script.
3. The macros engine has performance penalties when repeated over ~500-1000 widgets and should be avoided in favor of creating a class with args and loading a ui file exactly once. It's possible to optimize this on the pydm end but it's hard to do so without knock-on effects.
